### DISCUSSION TITLE
Import of a file in Unimarc format.

### DISCUSSION THREAD

**Discussion by meteos77** [textbrowser/biblioteq#113] [Q&A] [answered]
Hello, I'm coming to you because I'm a volunteer in a library (French) and I'm currently testing your software. I am currently using a software that has disappeared "papyrus" and I am looking for a new one.
I am looking for a new free software if possible.

A lack appears in your software for our own needs:
- Import of a file in unimarc format (with setting of fields 995 - exchange between libraries - maybe a French specificity ?)

I am not a c++ developer so I can't help you with this feature.

Do you think you can improve your software in this sense in a reasonable time ?

thank you in advance for your answer

Translated with www.DeepL.com/Translator (free version)

**Comment by textbrowser** [comment]
I will need a sample file to import. :) As for time required to complete the task, it varies with the complexity of the request and my availability. Thank you for the interest.

**Comment by meteos77** [comment]
thank you for your interest in my request
I am attaching a record file that can be accessed by the yaz-marcdump tool
[4_test-1_Notices-BDP.pan.txt](https://github.com/textbrowser/biblioteq/files/7452819/4_test-1_Notices-BDP.pan.txt)
the file extension has been modified with the addition of .txt for guthub

**Reply by meteos77** [reply]
I just looked at the sources and it looks like a z3950 import
void biblioteq_marc::parseBookZ3950Unimarc but from a file of x records
and some extra fields filled in the 995 field for inter-library exchange
best regard

**Comment by meteos77** [comment]
I realize that the work to create a graphical interface is hard and requires a lot of time.
I propose you a 1st version for the request of unimarc file import functionality
if you want to work in this direction, i can put in graphic with qtcreator my thoughts with a .ui file, and make the explanatory .odt of my request. you still have the hardest part - readapt my interface with the right parameters and code in c++ the backoffice

when do you think ?

[2021-11-04_unimarc_import_file-v1.tar.gz](https://github.com/textbrowser/biblioteq/files/7474782/2021-11-04_unimarc_import_file-v1.tar.gz)

**Comment by textbrowser** [comment]
It is not difficult, it requires a lot of time. I do not enjoy specific import tools. I would rather have a generic tool that can be customized for any input. The current one for books does that. It is incomplete because it handles single items, it does not allow for saving custom templates, and probably some other themes. Yes, please provide an example of what you'd like it to look like. :)

**Comment by meteos77** [comment]
I mean it that way (an integrated tool)
this is what I tried to do in the tar.gz
a .ui to show you what I expected in terms of presentation and an .odt for the explanations ???

see 1st line of the .odt
Importation
BiblioteQ offers a mechanism for importing book data contained in a Unimarc file. Please consult the Tools menu.
as in your doc

**Comment by textbrowser** [comment]
Aha! I only extracted a single file. Let me see.

**Comment by textbrowser** [comment]
OK. The focus will be on the first Unimarc Import tab. You'll need a delimiter. I'll have more questions when I begin. :)

**Comment by meteos77** [comment]
at your disposal; tomorrow I will try to finalize the 3 other tabs because they are linked and I would not like you to go in a wrong direction

have a nice evening ... well I don't know your time zone

**Comment by meteos77** [comment]
attention my .ui file is probably very bad it's only one week that I practice qtcreatorfile:///home/yo/Bureau/10-BiblioteQ/2021-11-05_unimarc_import_file-v2.tar.gz

**Comment by textbrowser** [comment]
Don't worry. I'll add the theme to the existing import tool and post images of my progress.

**Comment by meteos77** [comment]
I did not see import unimarc in your todo list :-o

**Comment by textbrowser** [comment]
The discussions are a pleasant substitute. :)

**Comment by meteos77** [comment]
I won't be able to talk to you because I know 30-50 words of English
on the other hand we can exchange with some difficulties when the words are too complex (double meaning ... too specific to the language) but on the whole it works quite well
I installed my pi with debian 11
I'm compiling yesterday's version

**Comment by textbrowser** [comment]
:)

https://github.com/textbrowser/biblioteq/commit/a67e2adb420e23cd0f6400b9848c171fa0f4b170

**Comment by meteos77** [comment]
for the library it is an essential feature

**Comment by meteos77** [comment]
je vous joints un fichier avec 9 notices unimarc
[3_test-9-Notices-unimarc_UTF-8-source-Bnf.pan.tar.gz](https://github.com/textbrowser/biblioteq/files/7513985/3_test-9-Notices-unimarc_UTF-8-source-Bnf.pan.tar.gz)

**Comment by meteos77** [comment]
for the standard unimarc I suppose it is like the z3950 but from a file 

but for the 995 fields

suggestion : create new fields in the database but generic fields so that other users can use them for other things.

for example : fieldspersotext1 , fieldspersotext2 , fieldsperso date1 ... the different types of data

on the other hand in the interface, create a panel with the personal label for each field

for example fieldspersotext1 , I want it to be named in the user interface "PERSONAL SOURCE
another user can name it in the user interface "MONTAGE"

**Reply by textbrowser** [reply]
Yes, generic fields. Please see https://github.com/textbrowser/biblioteq/blob/master/SQL/postgresql_create_schema.sql.

Then, submit a ticket and include the field name, field type, field attributes, and the name of the database table.

**Comment by meteos77** [comment]
The problem with these data (those in field 995) is that they 
generally coded

for example 
995 $j
$j
Section
Not mandatory, not repeatable, fixed length (1 character) This information concerns the section of the library for which the loaned document is intended. It is to be
It is to be distinguished from the information on the target audience expressed in subfield $q, which concerns the document itself
and not the set to which it is intended to be attached.
List of values to use :
a = adults
b = youth
c = teenagers
d = music
e = video
u = indeterminate
0 to 9 = determined sections within a network, provided that the table is published

**Comment by meteos77** [comment]
I am attaching recommendation 995 for your information.
[recommandation-995.pdf](https://github.com/textbrowser/biblioteq/files/7514084/recommandation-995.pdf)
some fields give info in the world standard unimarc fields others are completely local

**Reply by textbrowser** [reply]
Do you have an English version?

**Comment by meteos77** [comment]
no it is for the exchange of documents between French libraries

I will translate this document for you

on the other hand in your country, does it use the 9XX field
"9XX National and local data block

**Reply by textbrowser** [reply]
Aha, now we've entered the realm of personalized software; beyond preferences. Hmmm. Well, as long as the database field names are generic, things will be fine. The new import will be specialized and generic. Yah, all fine.

**Comment by meteos77** [comment]
I don't know if it's a translation problem but for me it's generic and customizable

**Comment by meteos77** [comment]
I still have to think about it but I'll give you my ideas
!! Don't develop anything until it's clear !

to be as generic as possible you should not even know the 995

that would allow to define something really generic

scenario: a Canadian wants to have a unimarc field taken into account while the standard import did not

by hypothesis: the 995 field does not exist in Canada so if we really want to be generic, it is up to the user to specify that there is an info that he wants to integrate in the database

so the import interface must ask which field he wants
for me it's 995$a
but the Canadian may want the 936$r

then there is the retrieval of the info and the transformation of this info into useful info (coded info)

for the retrieval of the data, you may need a regex because you won't know what format to expect

I take the example of the section 995$j => 

if 995 $j = a -> I want it to indicate adults in the database
if 995 $j = b -> I want it to indicate young in the database

so have an option that the user fills in a=adult ; b=young ...

and for the affection in the customizable fields

I want the section 995$j to be stored in the persotext1 field
if it was dates like the 995$m and $n => persodate1 and persodate2

**Comment by meteos77** [comment]
for me it's over for today
good thoughts on what to put in place

**Comment by meteos77** [comment]
I have translated the recommendation 995 with my comments

[translation_recommendation_995.pdf](https://github.com/textbrowser/biblioteq/files/7521073/translation_recommendation_995.pdf)

**Comment by textbrowser** [comment]
Thanks!

**Comment by meteos77** [comment]
On Saturday, I was at our library and our software crashed (restoration but 1 month of returned loans to reintegrate)

I have the possibility to have an appointment with the person in charge of the loans of the BDP on December 15th, do you think you will have implemented the import unimarc 995 by this date?
Even if I know that it is a big job, it will allow me to validate with her the management of the import of their file with your software.
good sunday

**Reply by textbrowser** [reply]
In a month? Most unlikely. :)

**Comment by meteos77** [comment]
it's going to be hard because I have to maintain the old one as much as possible which is going to take me some time while I prefer to work on the new one :-(

**Comment by textbrowser** [comment]
How did the software crash?

**Comment by meteos77** [comment]
it is an old paradox database, there are inconsistencies in the data entered because the fields do not have enough validation control

for example you can enter an isbn instead of the member number, you don't feel it (the shower head is good but the data is quickly entered in the wrong place :-) )

books that are already returned are no longer available for loan, probably a problem in the table of loans / returns / users

the people of the library don't know the software completely.
the database has never been maintained in 8 years (I think it has held up well)

a stroke of luck I had gone through on 7/10 and had taken a backup to work with the future software and I went through yesterday doing a csv export just 15 minutes before the crash.

it is necessary to retrace what was done between 7/10 and 13/11
ask the volunteers what they did to verify

finally it's time to change the software

Translated with www.DeepL.com/Translator (free version)

**Comment by textbrowser** [comment]
I see. Another reason for using PostgreSQL. :) SQLite does not offer backups.

**Comment by meteos77** [comment]
For the moment I'm trying to make the .csv as clean as possible, this will be used for the data recovery 
to reconcile the fields of the old database to BilblioteQ ...
I would like to lose as little information as possible

**Comment by meteos77** [comment]
Papyrus exports 63 fields
BiblioteQ imports 23 :-)
fortunately the 63 fields are not used :-)

**Comment by textbrowser** [comment]
The idea of CSV files remains foreign to me. It's very antiquated. SQL is standard. If I manage to translate PQ to Android, BQ will be able to present data on a tablet. :) It can now of course but only for SQLite. See the image.

https://raw.githubusercontent.com/textbrowser/biblioteq/gh-pages/images/biblioteq_11.jpg

That's on a tablet!

**Comment by textbrowser** [comment]
Redoing the interface for Android would have cost at least a full-time year. Now, it's just another release bundle.

**Comment by meteos77** [comment]
we are on windows for the library station and the csv is the export format of papyrus
I can process the csv with a bash while I don't know how to do it with SQL 
so yes there is better, more modern, more adapted but there are constraints

**Comment by meteos77** [comment, marked as answer]
you do what you want, it's your hobby :-)
I will also take advantage of the new booking process!
I will try to see with yaz-marcdump + bash if I can transform my unimarc files in csv to be able to import them without waiting because I am afraid of a new crash

**Reply by textbrowser** [reply]
I mean that the import will work with the CSV because that is a standard format. I was referring to the general use of comma-separated files. They are suitable for transferring data between projects but nothing else.
